onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -37,30 +42,115 @@ onnxruntime::NodeArg* AddCastNode(onnxruntime::Graph& graph,
   return new_arg;
 }
 
-static bool IsInputFloat16(const onnxruntime::Node& node) {
-  for (auto input : node.InputDefs()) {
-    if (input->Type() != nullptr &&
-        DataTypeImpl::TypeFromProto(*input->TypeAsProto()) == DataTypeImpl::GetTensorType<MLFloat16>() &&
-        !node.GetExecutionProviderType().empty()) {
-      return true;
-    }
+// check if the node has an fp16 input but was not able to be assigned an execution provider.
+// we will need to add a casts to/from fp32 around the node for it to be executed
+static bool NodeNeedsInputCastToFp32(const onnxruntime::Node& node) {
+  bool not_assigned = node.GetExecutionProviderType().empty();
+
+  if (not_assigned) {
+    const auto& input_defs = node.InputDefs();
+    bool has_fp16_input = std::any_of(input_defs.cbegin(), input_defs.cend(),
+                                      [](const NodeArg* input_def) {
+                                        return IsMLFloat16Tensor(*input_def);
+                                      });
+    return has_fp16_input;
   }
+
   return false;
 }
 
-static bool IsSingleInputNodeFloat16Node(const onnxruntime::Node& node) {
-  if (IsInputFloat16(node) && node.GetExecutionProviderType() == kCpuExecutionProvider) {
-    for (auto it = node.InputNodesBegin(); it != node.InputNodesEnd(); ++it) {
-      if (IsInputFloat16(*it))
-        return false;
-    }
-    for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
-      if (IsInputFloat16(*it))
-        return false;
-    }
-    return true;
+// Detect an isolated node that is able to process fp16 data but is between other nodes that have fp16 inputs
+// but will need a Cast inserted to enable them to run.
+//
+// Say we have 3 nodes in the middle of a graph that all have fp16 inputs.
+//
+// -> NodeA -> NodeB -> NodeC ->
+//
+// NodeA and NodeC have no kernel that can handle fp16 data (no execution provider assigned).
+//   e.g. 'Add' does not have an fp16 kernel
+// NodeB has a kernel that can process fp16 data (assigned to CPU EP).
+//
+// By default, we would insert Cast to/from fp32 around NodeA and NodeC as all operators have an fp32 kernel.
+//
+// i.e. -> CastToFp32 -> NodeA -> CastToFp16 -> NodeB -> CastToFp32 -> NodeC -> CastToFp16
+//
+// We can avoid the casts around NodeB if we also force that to run using fp32 data.
```

Comment:
Problem with mentioning example ops is that we add support for types when needed/requested so the comment gets stale and becomes confusing.

Just trying to maintain existing behaviour - which is not guaranteed to improve perf anyway. If B is an expensive op it may have been cheaper to do the Cast and run it using fp16. Forcing more ops to run in fp32 when the model specified them as fp16 seems even more questionable.

And all that said, given a lot more ops have fp16 support these days I'm not sure how often this logic actually kicks in.

Reply:
Makes sense.  We could just delete this code entirely but I guess it is safer to maintain the existing behavior.